jhonrocha/engineering-interivew-be
# Setup Environment and Authentication+Authorization

### What does this PR do?
This PR includes:
- Authentication: based on username and password
- Authorization: based on users roles
- Docker: using docker-compose to provide containers for a MySQL DB
 
### Any background context you want to provide?
Authentication is done using local strategy: username and password.
Authorization is done via JWT and it expects a Bearer Token on Header of requests.
 
### Where should the reviewer start / What requires special attention?
Check the following endpoints:
```
1. Authentication:
  - POST    /login: authenticate a user.

1. Users CRUD:
  - GET     /profile: get current user data. **Users** only.
  - POST    /users: create a new user on DB. Open endpoint.
  - GET     /users/:id: get a user by id. Get all if no id specified. **Admin** only.
  - DELETE  /users/:id: delete a user by id. **Admin** only.
```
  
### How should this be manually tested?
- Start docker on the host machine
- On the root folder run the following command to start the containers:
```
docker-compose up -d
```
- On your local install dependencies:
```
npm install
```
- Now start the project:
```
npm start
```
- The endpoints should be available on host machine throug the host:
```
http://localhost:8300
```
 
### Questions:
- Does this add new dependencies be installed?

Yes. It is required to run `npm install` again.

- Are any environmental dependencies or operational support steps (sql commands, config changes, etc) required?

Yes. It is required to have Docker installed to run the commands.
